api/src/infrastructure/kubernetes/infrastructure.rs

```diff
@@ -0,0 +1,894 @@
+            match service.container_type() {
+                ContainerType::ApplicationCompanion | ContainerType::ServiceCompanion => continue,
+                _ => {}
+            };
+
+            let image = service
+                .image()
+                .cloned()
+                .expect("Service should contain image");
+            let mut service_config = ServiceConfig::new(service.service_name().clone(), image);
+
+            if let Some(port) = service.port() {
+                service_config.set_port(port);
+            }
+
+            configs.push(service_config);
+        }
+
+        Ok(configs)
+    }
+
+    async fn get_logs(
+        &self,
+        app_name: &String,
+        service_name: &String,
+        from: &Option<DateTime<FixedOffset>>,
+        limit: usize,
+    ) -> Result<Option<Vec<(DateTime<FixedOffset>, String)>>, Error> {
+        let mut p = ListParams::default();
+        p.label_selector = Some(format!(
+            "{}={},{}={}",
+            APP_NAME_LABEL, app_name, SERVICE_NAME_LABEL, service_name,
+        ));
+        let pod = match Api::v1Pod(self.client())
+            .within(app_name)
+            .list(&p)
+            .await?
+            .into_iter()
+            .next()
+        {
+            Some(pod) => pod,
+            None => {
+                return Ok(None);
+            }
+        };
+
+        let mut p = LogParams::default();
+        p.timestamps = true;
+        p.since_seconds = from
+            .map(|from| {
+                from.timestamp()
+                    - pod
+                        .status
+                        .as_ref()
+                        .unwrap()
+                        .start_time
+                        .as_ref()
+                        .unwrap()
+                        .0
+                        .timestamp()
+            })
+            .filter(|since_seconds| since_seconds > &0);
+
+        let logs = Api::v1Pod(self.client())
+            .within(&app_name)
+            .log(&pod.metadata.name, &p)
+            .await?;
+
+        let logs = logs
+            .split("\n")
+            .enumerate()
+            // Unfortunately,  API does not support head (also like docker, cf. https://github.com/moby/moby/issues/13096)
+            // Until then we have to skip these log messages which is super slow…
+            .filter(move |(index, _)| index < &limit)
+            .filter(|(_, line)| !line.is_empty())
+            .map(|(_, line)| {
+                let mut iter = line.splitn(2, ' ').into_iter();
+                let timestamp = iter.next().expect(
+                    "This should never happen: kubernetes should return timestamps, separated by space",
+                );
+
+                let datetime =
+                    DateTime::parse_from_rfc3339(&timestamp).expect("Expecting a valid timestamp");
+
+                let mut log_line: String = iter.collect::<Vec<&str>>().join(" ");
+                log_line += "\n";
```

Comment:
I am not too familiar with the `String` impl of rust. Are mutable strings easily growable? If not, it might be worthwhile to concat the newline before joining into the string, seeing that this op is performed for all requested log lines.